백준 1406번 문제 순서도

rasangchul opened the pull request:
## 백준 1406번 문제의 순서도와 코드입니다.

 - 코드는 실수로 순서도 리뷰가 끝나지 않은 상태에서 작성된 코드입니다.
 - 삭제하기 아까워 일단 올린 후, 순서도 리뷰가 끝나면 다시 작성하도록 하겠습니다.
 - editor, editor2 파일이 있는데, editor2가 완성된 파일입니다. (시간 초과로 실패)
 - editor4가 문제를 맞춘 코드입니다.

### 어려웠던 부분 

- 처음 연결리스트를 사용할려 하였지만 어떻게 연결해야 해야할지 몰라 실패!
- editor4 에서 마지막에 콘솔로 찍을때 arr를 합칠때 reverse를 해줘야 한다는 것을 콘솔로 따로 찍어보고 알았습니다.

### 코드 실행 방법

- npm run prob 명령어 실행

###  [1406번 문제 링크](https://www.acmicpc.net/problem/1406)

```js
dmih
11
B
B
P x
L
B
B
B
P y
D
D
P z
```
 라고 입력을 받으면 

```js
yxz
```
처럼 출력되어야 함


### 백준 코드 

```js
const fs = require('fs');
const input = fs.readFileSync('/dev/stdin').toString().trim();
let start = input.split('\n')[0].split('');
let commands = input.split('\n').slice(2); 
let arr = [];

commands.forEach((cmd) => {
    if( cmd.startsWith('P') ) {
        const char = cmd.split(' ')[1];
        start.push(char)
    } else if (cmd === 'B') {
        if(start.length > 0) {
            start.pop();
        }
    } else if (cmd === 'L') {
        if(start.length > 0) {
            arr.push(start[start.length-1]);
            start.pop();
        }
    } else if (cmd === 'D') {
        if(arr.length > 0) {
            start.push(arr[arr.length-1]);
            arr.pop();
        }
    }
});
console.log(start.join('') + arr.reverse().join(''));
```

rasangchul committed:
모든 예제 통과 but 시간초과로 실패

rasangchul committed:
순서도 수정

Malloc72P commented:
처음 계획대로, 우선 순서도 리뷰가 끝나면 코드리뷰를 진행하겠습니다.

해당 PR은 일단은 close해주시거나, draft로 변경해주세요

rasangchul committed:
2번째 시도

rasangchul committed:
순서도 수정

rasangchul committed:
문제 성공적으로 맞춤

rasangchul commented:
### 내용

- editor 파일
> 처음 문제를 이해하기 위해 시도한 코드입니다. 보기에 더러울 수 있습니다.

- editor2 파일 
> 문제를 이해하고 풀어보았습니다. 하지만 시간 초과로 실패

- editor3 파일 
> 스택을 이용해볼려고 했습니다. 하지만 메모리 초과로 실패

- editor4 파일 
> 배열을 이용해 풀었습니다. 맞췄습니다.

### 수정된 부분 

- editor 3,4 가 추가 되었습니다.
- editor4의 순서도는 승철선배가 생략하라 하여 없습니다.

시간 나실때 리뷰 한번씩 해주세요.

Malloc72P commented:
준비된거면 리뷰어 등록 해주세요.

Malloc72P commented on problem/editor4.js:
```diff
@@ -0,0 +1,36 @@
+L
+B
+P y`;
+let start = input.split('\n')[0].split('');
+let commands = input.split('\n').slice(2); 
```
split이 두번 발생합니다.

split은 문자열의 처음부터 끝까지 순회할텐데요,
가능하면 한번만 호출하는게 좋을 것 같네요.

rasangchul replied:
제가 원하는 것이 입력의 첫번째 값을 배열로 만드는 것 입니다. 
그렇게 하기 위해서는 split이 두번쓰일 수 밖에 없었습니다. 
- 첫번째 split 입력값의 첫번째 부분 추출하기 위해 사용되었습니다.
- 두번째 split 입력값을 배렬로 만들기 위해 사용되었습니다.

한번만 호출할려면 이런 방법도 있겠네요

```js
const firstInput = input.split('\n')[0];
let start = firstInput.split('');
```
p.s start 변수명을 수정을 했습니다. 하지만 여기에선 그냥 start로 했습니다.

Malloc72P replied:
전달이 잘못된 것 같네요.

`input.split('\n')` 이게 두번 호출되어 문제라고 한 겁니다.
이미 `input`에 대해 `\n`을 딜리미터로 사용해서 `split`했는데, 이걸 또 할 필요가 있을까요?

rasangchul replied:
아하 그럼 이렇게 수정하면 되겠네요
```js
const inputLines = input.split('\n'); 
let start = inputLines[0].split(''); 
let commands = inputLines.slice(2); 
```

Malloc72P commented on problem/editor4.js:
```diff
@@ -0,0 +1,36 @@
+L
+B
+P y`;
+let start = input.split('\n')[0].split('');
+let commands = input.split('\n').slice(2); 
+let arr = [];
+
```
변수명만 봐선 어떤 목적인지 잘 알 수 없네요.

start라고 되어있는데 왜 이름이 start인가요?

또, arr인데, 왜 arr인가요?

변수명만 봐도 목적을 알 수 있으면 좋겠네요

Malloc72P replied:
또, 변수 선언시 전부 let을 사용하는데요, 왜 let을 사용했나요?

18번 라인을 보면 const를 사용하기도 하는데, let과 const의 차이를 설명할 수 있나요?

rasangchul replied:
const로 선언된 배열과 객체에 요소나 속성을 추가할 수 있는 이유

- `push` 와 `pop` 행위가 `재할당/재선언`이 아니기 때문입니다.

-----------------------------------------------------------

**`const` 와 `let` 의 차이점은**

#### `const`

- `재할당/재선언` 불가능

#### `let`

- 재할당 가능 (해당 블록 내에서만)
- 재선언 불가능

Malloc72P replied:
네, 둘 다 재선언이 불가능하다는 공통점이 있지만 재할당에서 차이가 발생합니다.

또, 가능하면 변수 선언 시 const를 사용하는게 좋습니다.

let 으로 선언하면 재할당이 가능하니, 어디선가 해당 변수의 값이 바뀔 수 도 있다는 걸 염두에 두면서 작업해야 합니다.
const로 선언되었다면, 재할당이 불가능할테니, 나중에 해당 변수를 사용할 때 할당된 값이 변경되지 않을거라는 보장이 생기기 때문입니다.

재할당이 발생하지 않는 변수라면 const를 사용해주세요.

Malloc72P commented on problem/editor4.js:
```diff
@@ -0,0 +1,36 @@
+    if( cmd.startsWith('P') ) {
+        const char = cmd.split(' ')[1];
+        start.push(char)
+    } else if (cmd === 'B') {
+        if(start.length > 0) {
+            start.pop();
+        }
```
해당 문제는 연결리스트로 풀면 시간초과가 나고, 스택으로 풀면 통과하는데 왜 그럴까요?

스택을 사용한 이유가 될텐데 한번 고민해보고 코멘트 달아주세요

rasangchul replied:
**연결리스트**

- 연결리스트에서 요소에 접근하기 위해서는 요소들을 하나씩 순회해야 합니다.
 - 때문에 원하는 부분에 요소를 삭제, 추가 하기위해서는 접근이 필요하고 순회해야하기에 속도가 느립니다.
 - 해당 노드에 접근하는 데 O(n)이 걸립니다.

**스택**

- 추가하거나 제거하는 연산은 일반적으로 상수 시간 O(1)입니다. 

때문에 이 문제를 푸는경우 연결리스트로 풀게되면 시간초과가 나게 되는것입니다.

Malloc72P replied:
#### 노드 이동의 경우

네, 단순 링크드 리스트의 경우 맨 마지막 노드에 접근하려면, 맨 처음 노드부터 끝까지 이동해야합니다.
물론 링크드 리스트 구현 방식에 따라, 맨 마지막 노드에 대한 참조를 가지는 경우도 있어서, 링크드리스트라고 해도 탐색 성능은 개선 가능합니다.

#### 스택의 특성

- 단순히 순회에 대해서만 생각해본 것 같은데요, 해당 문제의 정답이 스택인건, 스택의 특성때문입니다.
- 스택은 후입 선출이라는 특성을 가집니다. 맨 마지막에 넣은게 먼저 나옵니다.
- 커서를 우측으로 이동시키면, 커서 좌측을 표현하는 자료구조에 커서 우측의 문자를 넣어야 합니다.
- 커서를 좌측으로 이동시키면, 좌측 자료구조에서 맨 마지막에 넣은 문자를 꺼내서 우측 자료구조에 넣어야 합니다.
- 이러면 좌측 자료구조가 스택의 성질을 띄어야 하겠죠. 가장 마지막에 넣은게 먼저 나와야 하니까.
- 그래서 스택 자료구조를 사용해야 하는 문제였던거예요.

rasangchul replied:
네 알려주셔서 감사합니다.

Malloc72P reviewed (changes requested):
리뷰 완료했습니다. 코멘트 확인해주세요.

### editor 파일이 여러 개 존재하는데요,

- 커밋했다면 언제든지 이전 버전으로 돌릴 수 있습니다.
- git 기능을 잘 활용한다면 editor1, 2, 3, 4 이렇게 작성하지 않아도 됩니다.

### 주석이 없네요

- 왜 주석이 하나도 없나요?

KIMSANGYEOB reviewed (commented):
P1.

가장 기본적인 작업 중 하나가 줄정리입니다. 정리되지 않는 코드들을 정리해주세요.

리뷰하기 전 코드는 반드시 한번 더 점검하고 줄정리는 꼭 지켜서 올려주세요. 🫥

rasangchul committed:
줄 정리

rasangchul committed:
메소드 호출시 닫힌 중활호에 세미콜론 붙임

rasangchul committed:
줄 정리

- 수정이 안된 부분을 수정하였습니다.

rasangchul commented:
## 줄 정리 규칙

- 하나의 명령이 끝났을 때 세미콜론을 붙였습니다.
- 변수 선언, 할당할 경우 세미콜론을 붙였습니다.
- 배열의 값을 할당할 경우 세미콜론을 붙였습니다.
- 해당 블록에서 구문이 끝난 후 다른 명령어가 나올 때는 세미콜론을 붙였습니다.
- 함수나 조건문, 반복문의 중괄호 (if, for, while 등) 에서 닫힌 중괄호에 세미콜론을 삭제하였습니다.
- 메서드 호출에서 닫힌 중괄호에 세미콜론을 추가하였습니다.
- if문, for문 다음에 오는 괄호를 띄었습니다. 
  - ex) for (let i = 0; i < a; i++) { ...
- for문 다음에 for문이 오는 경우 줄 바꿈을 했습니다.
- if문 다음으로 오는 if문은 줄바꿈을 하지 않았습니다. 
  - if문은 이어지는 경우가 많기 때문입니다.
- 변수 선언 다음으로 for문, if문 이 올 경우 줄 바꿈을 했습니다.

## 변수명 수정

- start -> beforeCursor 
  - 커서 앞의 문자들을 저장하므로 beforeCursor 로 수정하였습니다.
- arr -> afterCursor
  - 커서 뒤의 문자들을 저장하므로 afterCursor 로 수정하였습니다.

## split이 두번 발생합니다.

- 첫번째 split은 입력값의 첫번째값(abc)를 추출하기 위해 사용했습니다.
- 두번째 split은 추출한 값(abc)을 나눠 배열로 만들기 위해 사용했습니다.
- 왜 배열로 만들려고 했나요?
  - 문자를 빼고, 넣기위해 배열로 만들었습니다.

## 해당 문제는 연결리스트로 풀면 시간초과가 나고, 스택으로 풀면 통과하는데 왜 그럴까요?

- 이 부분은 이슈로 올리겠습니다.

##  변수 선언시 전부 let을 사용하는데요, 왜 let을 사용했나요?

- 이부분은 이슈로 올리겠습니다.

rasangchul committed:
변수명 수정

- textBeforeCursor가 너무 길어 beforeCursor로 수정하였습니다.
- textAfterCursor가 너무 길어 afterCursor로 수정하였습니다.

Malloc72P commented:
뭔가 전달이 잘못된 것 같은데...

리뷰 수정사항 반영하고 나서 코멘트 올려달라고 한건, conversation에 답글을 안달아도 된다는 의미가 아닙니다.

Malloc72P commented:
이 부분은 이슈로 올리겠다고 하셨는데요,

이슈로 올려야만 하는 이유가 있을까요?
해당 pr에서 답변할 수 없을 정도로 내용이 긴가요?

Malloc72P commented:
줄 정리 규칙의 경우...

그냥 prettier 설정 만들고 자동 포맷 돌리면 편할텐데, 수작업으로 하는 이유가 있나요?

rasangchul commented:
- conversation 에 하나하나 답글을 다는것 보단 한번에 정리해서 코멘트로 올리는게 보기 편할거라고 생각해 이렇게 했습니다.
   - 보기 불편하시면 conversation답글 다는 것으로 수정하도록 하겠습니다.
- 이슈로 따로 만들어서 올리려는 이유는, 해당 내용을 정리해두면 나중에 필요한 시점에 쉽게 찾을 수 있고, 실무나 알고리즘 문제를 풀 때도 참고 자료로 활용할 수 있다고 생각해서 이슈로 올릴려고 했습니다.
- 줄 정리 같은 경우 prettier 쓰면 되는걸 알고 있지만 
  - 수작업으로 정리하다 보면 코드 스타일에 대해 더 잘 이해하게 되고, 저만의 패턴도 배울 수 있어서 수작업을 통해 배우는 과정도 중요하다고 생각해서 수작업으로 했습니다.

Malloc72P commented:
> * 이슈로 따로 만들어서 올리려는 이유는, 해당 내용을 정리해두면 나중에 필요한 시점에 쉽게 찾을 수 있고, 실무나 알고리즘 문제를 풀 때도 참고 자료로 활용할 수 있다고 생각해서 이슈로 올릴려고 했습니다.

- 이슈는 해야할 일을 표현하기 위한 도구입니다.
- 정리를 위한거라면 개인 노트에 해주세요.
- 만약 공유가 필요한 내용이라면, wiki를 활용해봐도 좋겠습니다.

> 줄 정리 같은 경우 prettier 쓰면 되는걸 알고 있지만

- 우리캠프 시작할 때, 필요하면 eslint와 prettier를 직접 구성해서 사용하라고 했었으니, 그냥 prettier를 사용하는것을 권장합니다.

> 수작업으로 정리하다 보면 코드 스타일에 대해 더 잘 이해하게 되고, 저만의 패턴도 배울 수 있어서 수작업을 통해 배우는 과정도 중요하다고 생각해서 수작업으로 했습니다.

- 명확한 코딩 컨벤션이 있다면, 맞는 방법일 수 도 있겠습니다만, 따로 정한 컨벤션이 없는 상황입니다.
- 어떤 기준에 맞춰서 코드 스타일을 지정해야 하겠으나, 지금 그게 없으니, 그냥 prettier를 사용하라는 것 입니다.
- prettier로 정리하면, prettier 기본 컨벤션에 맞게 정리되니, 우선은 prettier가 어떻게 코드를 정리하는지 부터 보는게 좋겠습니다.

rasangchul commented:
네 알겠습니다.

rasangchul committed:
prettier 적용

- prettier 를 설치해 저장시 포맷이 적용되게 하였습니다.

rasangchul committed:
Remove node_modules from Git tracking

rasangchul committed:
.gitinore 파일 추가

- node_modules폴더를 git에 올리지 않기 위한 설정입니다.

rasangchul committed:
node_modules 삭제

rasangchul committed:
.gitinore 설정

- node_modules를 github에 올리지 않기 위한 설정입니다.

rasangchul committed:
Remove node_modules from tracking

rasangchul committed:
.gitignore 파일 만들기

rasangchul committed:
.gitignore 마지막 시도

rasangchul committed:
.gitignore 설정

rasangchul committed:
.gitignore 내용 작성

- DS_Store(macOS 운영체제에서 Finder로 폴더를 볼 때마다 자동으로 생성되는 폴더) 제외
- node_modules폴더 제외

rasangchul committed:
write .gitignore

rasangchul committed:
주석 추가

Malloc72P commented on problem/editor4.js:
```diff
@@ -0,0 +1,39 @@
+    if (cmd.startsWith("P")) {
+        const char = cmd.split(" ")[1]; // P로 시작할 경우 2번째값을 저장
+        beforeCursor.push(char);
+    } else if (cmd === "B") {
+        if (beforeCursor.length > 0) {
+            beforeCursor.pop();
+        }
+    } else if (cmd === "L") {
+        if (beforeCursor.length > 0) {
+            afterCursor.push(beforeCursor[beforeCursor.length - 1]);
+            beforeCursor.pop();
+        }
+    } else if (cmd === "D") {
+        if (afterCursor.length > 0) {
+            beforeCursor.push(afterCursor[afterCursor.length - 1]);
+            afterCursor.pop();
+        }
+    }
+});
+
+console.log(beforeCursor.join("") + afterCursor.reverse().join(""));
```
왜 P 분기 외에는 주석이 없나요?

별로 중요하지 않은 로직인가요?

rasangchul replied:
필요한 부분에 주석 다시 달아서 수정했습니다.

Malloc72P commented on problem/editor4.js:
```diff
@@ -0,0 +1,39 @@
+let commands = inputLines.slice(2); // 세 번째부터 끝까지 배열로 담기
+let afterCursor = [];
+
+commands.forEach((cmd) => {
+    if (cmd.startsWith("P")) {
+        const char = cmd.split(" ")[1]; // P로 시작할 경우 2번째값을 저장
+        beforeCursor.push(char);
+    } else if (cmd === "B") {
```
오프라인 리뷰에서 나왔던 것 처럼, 각 명령어가 뭐 하는 명령어인지도 적어주면 좋겠습니다.

또, P 명령어의 경우에도 어떤 일을 수행하는지 자세한 주석이 없네요?

rasangchul replied:
해당 문제를 이해하기 위해 사이트를 방문하지 않아도 이해할 수 있을만큼 설명을 주석으로 썼습니다.

Malloc72P reviewed (changes requested):
리뷰 완료했습니다. 코멘트 확인해주세요.

보니까 main브랜치에 이미 node_modules가 있었네요.

이번 pr에서 node_modules가 지워지면서, 리뷰하기 매우 불편해졌습니다.
![image](https://github.com/user-attachments/assets/b8c5910c-bb53-4ff7-8c34-f1dedad201e2)
리뷰 외에도 문제가 많은 일이니, 다음부터는 주의해주세요.

rasangchul committed:
주석 추가

KIMSANGYEOB reviewed (approved):
🆗
